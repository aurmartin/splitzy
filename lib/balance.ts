import dinero, { type Dinero } from "dinero.js";
import { type Expense, ExpenseService, useExpenses } from "@/lib/expenses";
import { type Group } from "@/lib/groups";

export type Balance = Record<string, Dinero>;

export type Reimbursement = {
  from: string;
  to: string;
  amount: Dinero;
};

export function computeBalance(group: Group, expenses: Expense[]): Balance {
  const balance = group.members.reduce((acc, member) => {
    acc[member] = dinero({ amount: 0, currency: group.currency });
    return acc;
  }, {} as Balance);

  expenses.forEach((expense) => {
    const payer = expense.payerName;
    const total = expense.splitExpense.total;

    balance[payer] = balance[payer].add(total);

    const amounts = ExpenseService.getAmounts(expense);

    Object.entries(amounts).forEach(([name, amount]) => {
      balance[name] = balance[name].subtract(amount);
    });
  });

  return balance;
}

export const useBalance = (group: Group): Balance => {
  const expenses = useExpenses(group.id);
  return computeBalance(group, expenses);
};

export function computeOptimalReimbursements(
  balance: Balance,
): Reimbursement[] {
  const reimbursements: Reimbursement[] = [];
  const members = Object.keys(balance);

  // Create a copy of the balance object
  const workingBalance = members.reduce((acc, member) => {
    acc[member] = balance[member];
    return acc;
  }, {} as Balance);

  // Continue while there are non-zero balances
  while (members.some((member) => !workingBalance[member].isZero())) {
    // Find the person who owes the most (most negative balance)
    const maxDebtor = members.reduce((a, b) =>
      workingBalance[a].lessThan(workingBalance[b]) ? a : b,
    );

    // Find the person who is owed the most (most positive balance)
    const maxCreditor = members.reduce((a, b) =>
      workingBalance[a].greaterThan(workingBalance[b]) ? a : b,
    );

    // If both balances are effectively zero (less than 1 cent), break
    if (
      workingBalance[maxDebtor].isZero() &&
      workingBalance[maxCreditor].isZero()
    ) {
      break;
    }

    // Calculate the reimbursement amount (minimum of absolute values)
    const debtorAmount = workingBalance[maxDebtor].multiply(-1); // Make positive
    const creditorAmount = workingBalance[maxCreditor];
    const amount = debtorAmount.lessThan(creditorAmount)
      ? debtorAmount
      : creditorAmount;

    // Update the balances
    workingBalance[maxDebtor] = workingBalance[maxDebtor].add(amount);
    workingBalance[maxCreditor] = workingBalance[maxCreditor].subtract(amount);

    // Record the reimbursement
    reimbursements.push({
      from: maxDebtor,
      to: maxCreditor,
      amount,
    });
  }

  return reimbursements;
}
