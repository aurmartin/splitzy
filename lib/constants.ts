export const primaryHue = 200;

export const Colors = {
  primaryDark: `hsl(${primaryHue} 80% 40%)`,
  primary: `hsl(${primaryHue} 40% 50%)`,
  primaryLight: `hsl(${primaryHue} 40% 80%)`,

  secondary: `hsl(0 0% 70%)`,
  secondaryLight: `hsl(0 0% 80%)`,

  red: `hsl(0 50% 40%)`,

  green: `hsl(120 50% 40%)`,

  background: `hsl(0 0% 94%)`,
};
