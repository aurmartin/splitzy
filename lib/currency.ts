import { type Dinero, type Currency } from "dinero.js";

const MINOR_UNIT_FACTOR: Partial<Record<Currency, number>> = {
  EUR: 100,
  USD: 100,
};

const getAmountColor = (amount: Dinero): string => {
  if (amount.getAmount() > 0) return "hsl(120 70% 30%)";
  if (amount.getAmount() < 0) return "hsl(0 70% 30%)";
  return "black";
};

const getAbsoluteAmount = (amount: Dinero): Dinero => {
  return amount.getAmount() < 0 ? amount.multiply(-1) : amount;
};

const getBalanceMessage = (amount: Dinero): string => {
  if (amount.getAmount() > 0) return "On vous doit";
  if (amount.getAmount() < 0) return "Vous devez";
  return "Vous êtes à jour";
};

const getMinorUnitAmount = (amount: number, currency: Currency): number => {
  const factor = MINOR_UNIT_FACTOR[currency];
  if (!factor) {
    throw new Error(`Unsupported currency: ${currency}`);
  }
  return amount * factor;
};

const getMajorUnitAmount = (amount: number, currency: Currency): number => {
  const factor = MINOR_UNIT_FACTOR[currency];
  if (!factor) {
    throw new Error(`Unsupported currency: ${currency}`);
  }
  return amount / factor;
};

export {
  getAmountColor,
  getAbsoluteAmount,
  getBalanceMessage,
  getMinorUnitAmount,
  getMajorUnitAmount,
};
