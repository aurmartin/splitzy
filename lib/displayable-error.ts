export class DisplayableError extends Error {
  title: string;
  message: string;

  constructor(title: string, message: string) {
    super(`${title}: ${message}`);
    this.title = title;
    this.message = message;
  }
}
