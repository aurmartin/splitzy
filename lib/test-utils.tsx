import { SystemProvider } from "@/components/system-provider";
import { Database, tables } from "@/lib/db/schema";
import { render, type RenderOptions } from "@testing-library/react-native";
import BetterSqliteDatabase from "better-sqlite3";
import { drizzle } from "drizzle-orm/better-sqlite3";
import {
  RenderRouterOptions,
  renderRouter,
  type MockContextConfig,
} from "expo-router/testing-library";
import { HttpResponse, http } from "msw";
import { setupServer } from "msw/node";
import React, { type ReactElement } from "react";
import { System } from "./system";
import * as jws from "jws";

const AllTheProviders = ({
  children,
  system,
}: {
  children: ReactElement;
  system: System;
}) => {
  return <SystemProvider system={system}>{children}</SystemProvider>;
};

const customRender = (
  ui: ReactElement,
  system: System,
  options?: Omit<RenderOptions, "wrapper">,
) => {
  const wrapper = ({ children }: { children: ReactElement }) => (
    <AllTheProviders system={system}>{children}</AllTheProviders>
  );

  return render(ui, { wrapper, ...options });
};

const customRenderRouter = (
  context: MockContextConfig,
  system: System,
  options?: Omit<RenderRouterOptions, "wrapper">,
) => {
  const wrapper = ({ children }: { children: ReactElement }) => (
    <AllTheProviders system={system}>{children}</AllTheProviders>
  );

  return renderRouter(context, { ...options, wrapper });
};

const createDatabase = () => {
  const db = new BetterSqliteDatabase(":memory:");
  return drizzle(db);
};

const clearDatabase = (db: Database) => {
  db.delete(tables.groups).run();
  db.delete(tables.expenses).run();
};

const createSupabaseServer = () =>
  setupServer(
    http.get("http://localhost:50001/rest/v1/groups", () =>
      HttpResponse.json([]),
    ),
    http.get("http://localhost:50001/rest/v1/expenses", () =>
      HttpResponse.json([]),
    ),
    http.post("http://localhost:50001/rest/v1/expenses", () =>
      HttpResponse.json({}),
    ),
    http.patch("http://localhost:50001/rest/v1/expenses", () =>
      HttpResponse.json({}),
    ),
    http.delete("http://localhost:50001/rest/v1/expenses", () =>
      HttpResponse.json({}),
    ),
    http.post("http://localhost:50001/auth/v1/otp", () =>
      HttpResponse.json({}),
    ),
    http.get("http://localhost:50001/auth/v1/user", () =>
      HttpResponse.json({
        id: "test-user-id",
        email: "test@test.com",
        confirmation_sent_at: "2025-03-01T00:00:00Z",
        created_at: "2025-03-01T00:00:00Z",
        updated_at: "2025-03-01T00:00:00Z",
      }),
    ),
  );

const setFakeSession = (system: System) => {
  const access_token = generateAccessToken();
  const refresh_token = "fake-refresh-token";

  system.setSession({ access_token, refresh_token });
};

const generateAccessToken = () => {
  const nowTimestamp = Math.floor(Date.now() / 1000);
  const expiresAt = nowTimestamp + 60 * 60 * 24 * 30; // 30 days

  const payload = {
    exp: expiresAt,
    iat: nowTimestamp,
    email: "test@test.com",
  };

  return jws.sign({
    header: { alg: "HS256" },
    payload,
    secret: "secret",
  });
};

export * from "@testing-library/react-native";
export {
  clearDatabase,
  createDatabase,
  createSupabaseServer,
  customRender as render,
  customRenderRouter as renderRouter,
  setFakeSession,
  generateAccessToken,
};
