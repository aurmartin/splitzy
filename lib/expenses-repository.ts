import { useSystem } from "@/components/system-provider";
import { expensesTable, type ExpenseRecord } from "@/lib/db/schema";
import { Receipt, type Expense, type Split } from "@/lib/expenses";
import { System } from "@/lib/system";
import dinero, { type Currency, type Dinero } from "dinero.js";
import { desc, eq } from "drizzle-orm";
import { useLiveQuery } from "drizzle-orm/expo-sqlite";

type DineroRecord = {
  amount: number;
  currency: Currency;
  precision: number;
};

const getExpense = (system: System, id: string): Expense | null => {
  const rows = system.db
    .select()
    .from(expensesTable)
    .where(eq(expensesTable.id, id))
    .limit(1)
    .all();

  if (rows.length === 0) {
    return null;
  }

  return decodeExpenseRecord(rows[0]);
};

const insertExpense = async (system: System, expense: Expense) => {
  const row = encodeExpense(expense);
  await system.syncEngine.insert(expensesTable, expense.id, row);
};

const updateExpense = async (system: System, expense: Expense) => {
  const row = encodeExpense(expense);
  await system.syncEngine.update(expensesTable, expense.id, row);
};

const deleteExpense = async (system: System, id: string) => {
  await system.syncEngine.delete(expensesTable, id);
};

const useExpenses = (groupId: string) => {
  const system = useSystem();
  const { data } = useLiveQuery(
    system.db
      .select()
      .from(expensesTable)
      .where(eq(expensesTable.groupId, groupId))
      .orderBy(desc(expensesTable.createdAt)),
  );

  return data.map(decodeExpenseRecord);
};

const useExpense = (id: string): Expense | null => {
  const system = useSystem();
  const { data } = useLiveQuery(
    system.db.select().from(expensesTable).where(eq(expensesTable.id, id)),
  );

  if (data.length === 0) {
    return null;
  }

  return decodeExpenseRecord(data[0]);
};

const decodeExpenseRecord = (record: ExpenseRecord): Expense => {
  return {
    id: record.id,
    groupId: record.groupId as string,
    title: record.title as string,
    payerName: record.payerName as string,
    splitExpense: decodeSplitExpense(record.splitExpense as string),
    receipt: decodeReceipt(record.receipt as string),
    createdAt: new Date(record.createdAt as string),
    updatedAt: new Date(record.updatedAt as string),
    deletedAt: record.deletedAt
      ? new Date(record.deletedAt as string)
      : undefined,
  };
};

const decodeSplitExpense = (encoded: string): Expense["splitExpense"] => {
  const decoded = JSON.parse(encoded);

  const total = dinero({
    amount: decoded.total.amount,
    currency: decoded.total.currency,
    precision: decoded.total.precision,
  });

  const members = decoded.members as string[];
  const type = decoded.type as Split["type"];

  switch (type) {
    case "percentage":
      return { type, total, members, ratios: decoded.ratios };
    case "equal":
      return { type, total, members };
    case "amount":
      const amounts = decoded.amounts as Record<string, DineroRecord>;
      return {
        type,
        total,
        members,
        amounts: Object.entries(amounts).reduce(
          (acc, [member, amount]) => {
            acc[member] = dinero({
              amount: amount.amount,
              currency: amount.currency,
              precision: amount.precision,
            });

            return acc;
          },
          {} as { [member: string]: Dinero },
        ),
      };
    case "receipt":
      return { type, total, members, receipt: decoded.receipt };
  }
};

const decodeReceipt = (encoded: string): Receipt | null => {
  const decoded = JSON.parse(encoded) as {
    date: string;
    total: DineroRecord;
    title: string;
    currency: Currency;
    items: {
      description: string;
      humanReadableDescription: string;
      price: DineroRecord;
      paid_for: string[];
    }[];
  } | null;

  if (decoded === null) {
    return null;
  }

  return {
    title: decoded.title,
    currency: decoded.currency,
    date: new Date(decoded.date),
    total: dinero({
      amount: decoded.total.amount,
      currency: decoded.total.currency,
      precision: decoded.total.precision,
    }),
    items: decoded.items.map((item) => ({
      description: item.description,
      humanReadableDescription: item.humanReadableDescription,
      price: dinero({
        amount: item.price.amount,
        currency: item.price.currency,
        precision: item.price.precision,
      }),
      paid_for: item.paid_for,
    })),
  };
};

const encodeExpense = (expense: Expense): ExpenseRecord => {
  return {
    ...expense,
    splitExpense: JSON.stringify(expense.splitExpense),
    receipt: expense.receipt ? JSON.stringify(expense.receipt) : null,
    createdAt: expense.createdAt.toISOString(),
    updatedAt: expense.updatedAt.toISOString(),
    deletedAt: expense.deletedAt?.toISOString() ?? null,
  };
};

export const ExpensesRepository = {
  getExpense,
  insertExpense,
  updateExpense,
  deleteExpense,

  useExpenses,
  useExpense,
};
