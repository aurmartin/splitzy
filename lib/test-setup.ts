import { System } from "@/lib/system";
import {
  clearDatabase,
  createDatabase,
  createSupabaseServer,
} from "@/lib/test-utils";

const server = createSupabaseServer();
let system: System;

beforeEach(async () => {
  clearDatabase(system.db);
});

beforeAll(async () => {
  server.listen();
  system = new System(createDatabase());
  await system.initializeMigrations();
});

afterEach(() => {
  server.resetHandlers();
});

afterAll(() => {
  server.close();
  system.dispose();
});

export { server, system };
