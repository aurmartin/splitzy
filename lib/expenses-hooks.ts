import { useSystem } from "@/components/system-provider";
import { Expense } from "@/lib/expenses";
import { ExpensesRepository } from "@/lib/expenses-repository";
import { useEffect, useState, useMemo } from "react";

function useExpenses(groupId: string): Expense[] {
  const system = useSystem();
  const initialExpenses = useMemo(
    () => ExpensesRepository.getGroupExpenses(system, groupId),
    [system, groupId],
  );
  const [expenses, setExpenses] = useState<Expense[]>(initialExpenses);

  useEffect(() => {
    return ExpensesRepository.watchGroupExpenses(system, groupId, setExpenses);
  }, [system, groupId]);

  return expenses;
}

function useExpense(id: string): Expense | null {
  const system = useSystem();
  const initialExpense = useMemo(
    () => ExpensesRepository.getExpense(system, id),
    [system, id],
  );
  const [expense, setExpense] = useState<Expense | null>(initialExpense);

  useEffect(() => {
    return ExpensesRepository.watchExpense(system, id, setExpense);
  }, [system, id]);

  return expense;
}

export { useExpense, useExpenses };
