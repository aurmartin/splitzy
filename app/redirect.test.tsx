import RedirectScreen from "@/app/redirect";
import { system } from "@/lib/test-setup";
import { generateAccessToken, renderRouter } from "@/lib/test-utils";
import { screen, waitFor } from "@testing-library/react-native";
import * as Linking from "expo-linking";
import { Text } from "react-native";

const routerContext = {
  "/redirect": () => <RedirectScreen />,
  "/login": () => <Text>Login Page</Text>,
  "/protected": () => <Text>Protected</Text>,
};

jest.mock("expo-linking", () => ({
  ...jest.requireActual("expo-linking"),
  useLinkingURL: jest.fn().mockReturnValue("/redirect"),
}));

describe("RedirectScreen", () => {
  it("should render snapshot", () => {
    renderRouter(routerContext, system, { initialUrl: "/redirect" });
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("should redirect to protected route when correct link is provided", async () => {
    const url = `/redirect?access_token=${generateAccessToken()}&refresh_token=test-refresh-token`;
    jest.mocked(Linking.useLinkingURL).mockReturnValue(url);
    renderRouter(routerContext, system, { initialUrl: url });
    await waitFor(() => expect(screen.getByText("Protected")));
  });

  it("should redirect to login page when no access_token is provided", async () => {
    const url = `/redirect?refresh_token=test-refresh-token`;
    jest.mocked(Linking.useLinkingURL).mockReturnValue(url);
    renderRouter(routerContext, system, { initialUrl: url });
    await waitFor(() => expect(screen.getByText("Login Page")));
  });

  it("should redirect to login page when errorCode is provided", async () => {
    const url = `/redirect?errorCode=test-error-code`;
    jest.mocked(Linking.useLinkingURL).mockReturnValue(url);
    renderRouter(routerContext, system, { initialUrl: url });
    await waitFor(() => expect(screen.getByText("Login Page")));
  });
});
