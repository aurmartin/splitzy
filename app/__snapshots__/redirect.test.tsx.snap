// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RedirectScreen should render snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNCSafeAreaView
    edges={
      {
        "bottom": "additive",
        "left": "additive",
        "right": "additive",
        "top": "additive",
      }
    }
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNCSafeAreaView
      edges={
        {
          "bottom": "additive",
          "left": "additive",
          "right": "additive",
          "top": "additive",
        }
      }
      style={
        [
          {
            "backgroundColor": "hsl(0 0% 94%)",
            "flex": 1,
            "padding": 8,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "gap": 16,
            "justifyContent": "center",
          }
        }
      >
        <ActivityIndicator
          color="hsl(200 40% 50%)"
          size="large"
        />
        <Text
          style={
            [
              {
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
              },
            ]
          }
        >
          Logging in...
        </Text>
      </View>
    </RNCSafeAreaView>
  </RNCSafeAreaView>
</RNCSafeAreaProvider>
`;
